runtime/compiler/runtime/CRRuntime.cpp

```diff
@@ -148,6 +157,177 @@ TR::CRRuntime::waitOnCRRuntimeMonitor()
    _crRuntimeMonitor->wait();
    }
 
+void
+TR::CRRuntime::pushMemoizedCompilation(TR_MemoizedCompilations& list, J9Method *method)
+   {
+   auto newEntry = new (_compInfo->persistentMemory()) TR_MemoizedComp(method);
+   if (newEntry)
+      list.add(newEntry);
+   }
+
+J9Method *
+TR::CRRuntime::popMemoizedCompilation(TR_MemoizedCompilations& list)
+   {
+   J9Method *method = NULL;
+   auto memComp = list.pop();
+   if (memComp)
+      {
+      method = memComp->getMethod();
+      jitPersistentFree(memComp);
+      }
+   return method;
+   }
+
+void
+TR::CRRuntime::removeMemoizedCompilation(TR_MemoizedCompilations& list, J9Method *method)
+   {
+   auto first = list.getFirst();
+   if (!first)
+      {
+      return;
+      }
+   else if (first->getMethod() == method)
+      {
+      popMemoizedCompilation(list);
+      }
+   else
+      {
+      auto prev = first;
+      auto curr = prev->getNext();
+
+      while (curr)
+         {
+         if (curr->getMethod() == method)
+            {
+            list.removeAfter(prev, curr);
+            jitPersistentFree(curr);
+            break;
+            }
+         prev = curr;
+         curr = curr->getNext();
+         }
+      }
+   }
+
+void
+TR::CRRuntime::removeMethodsFromMemoizedCompilations(TR_J9VMBase *fej9, TR_OpaqueClassBlock *clazz)
+   {
+   OMR::CriticalSection removeMemoizedCompilations(getCompilationMonitor());
+
+   J9Method *methods = static_cast<J9Method *>(fej9->getMethods(clazz));
+   uint32_t numMethods = fej9->getNumMethods(clazz);
+   for (uint32_t index = 0; index < numMethods ; index++)
```

Comment:
Would it be faster if we scan the list only once and for each entry we compare the class of the method to the class that gets unloaded?